[MRG] Format signature

smu095 opened the pull request:
A few initial questions concerning `format_signature` and the development workflow (apologies for the basic nature of these queries, thanks for being patient). In issue #8 you write 

> In general I think this task amounts to using the `inspect` library properly, in particular `inspect.Signature`. Then a few unit tests on user-defined functions and builtins like `list.index` must be added. Finally you can add the newly created function to the `doctree` output, run `doctree pandas`, `doctree collections`, etc and see if the output is as expected. If not, patch it up the code and add some more tests.

1. First things first, do you want `format_signature` to be a method of the `SimplePrinter` class?
2. Could you elaborate on the following: *"[t]hen a few unit tests on user-defined functions and builtins like `list.index` must be added"*? Assume that I am familiar with the contents of [this blog post](https://jeffknupp.com/blog/2013/12/09/improve-your-python-understanding-unit-testing/) (if you have any other useful beginner-level resources, please send them my way).
3. Concerning the development process: After adding some functionality to a package, does it need to be "recompiled" in some way, or can I use the new functionality right off the bat by calling `treedoc` (you write `doctree`, which I am assuming is a typo) with the appropriate arguments from the terminal?
4. In issue #8 I wrote: *"Do you have any useful resources/thoughts on type-checking and assertion statements for a task like this?"* Specifically, what I mean is that you are very diligent in checking e.g. that the inputs to a function satisfy a given number of conditions related to type, etc. I guess my question is: What is the guiding philosophy when it comes to checking that the input satisfies a certain set of conditions? Should one always, no matter how trivial, check that the input is what we expect it to be?

Thanks again!

tommyod commented:
Some of the answers are a bit short. I'm willing to discuss details, but it might make more sense to do it as you work on specific code related to the questions.

1. It will probably find better use as a function, i.e. not a method bound to a specific printer. Makes it easier to use it across printers.

2. To test on built-ins, you could (for instance) write tests like this:
```python
def test_format_signature():
   import collections
   assert format_signature(collections.Counter.most_common, 0) == ""
   assert format_signature(collections.Counter.most_common, 1) == "(...)"
   assert format_signature(collections.Counter.most_common, 2) == "(self, n)"
   assert format_signature(collections.Counter.most_common, 3) == "(self, n=None)"
```
You could also write a `def my_func(a, b=2)` for the sole purpose of testing on it. And we will parameterize the tests. In PR #6 there are some examples of tests you can look at. **If you approve that PR we can merge it into the master branch, and you can merge it into this branch.**

3. Good question! It's important that you install treedoc with `pip install -e .` in the treedoc directory where the `setup.py` file is located. This installs the package in editable mode. See `pip install --help` for documentation about the `-e` flag.

4. It's more important to check the input types at the higher levels of the API where users interact directly with the package. This function is so deep down in the structure that it's not too important. Besides, this function will accept a general python object: could be a function, a method, a class, a module, etc. So it's hard to sanitize the inputs.

tommyod commented on treedoc/utils.py:
```diff
@@ -82,6 +82,54 @@ def ispackage(obj):
+def format_signature(obj, verbosity=2):
+    # TODO: Figure out how to handle *args, **kwargs and *
+    
+    assert verbosity in (0, 1, 2, 3, 4)
```
You can write `assert 0 <= verbosity <= 4` if you prefer.

smu095 commented:
Ready for an initial review. In this PR:
* Added function `format_signature` to prepare signature for printing.
* Added `test_signature` to `test_utils.py`

Some questions:
1. I resolved some conflicts with master and merged into this branch (`0f17b02`). Should I be resolving conflicts at this point? What is considered best practice in terms of the order of the code review?
2. How can I see which checks are failing?

tommyod commented:
1. Yes, your branch should continually be resolved against the master branch. It's best to do it as soon as there's conflict.
2. Click on "Details", then on "The Build", or go to [PRs](https://travis-ci.com/tommyod/treedoc/pull_requests) on Travis. Open either the Python 3.6 or 3.7 job and scroll down. E.g. https://travis-ci.com/tommyod/treedoc/jobs/198395008.

tommyod commented on treedoc/tests/test_utils.py:
```diff
@@ -164,6 +165,34 @@ def test_resolve_object():
+    Test that formatting signature works.
+    '''
+    
+    def myfunc(a: float=4.2, b: int=42, *args, **kwargs):
```
Comment:Having this order of `*args` and `**kwargs` makes little sense, although syntactically correct. Positional argument follows keyword argument, so you would do `func(arg1, arg2, arg3, *args, kwarg1=1, kwarg2=3, kwarg3=5, **kwrags)` instead.

The test case `def myfunc(a, b:int, *args, c=4.2, d:int=42, **kwargs)` captures the more common usage, as well as a mix of type annotations.

tommyod commented on treedoc/utils.py:
```diff
@@ -289,6 +289,71 @@ def ispackage(obj):
 
+def _get_name(param):
+    ''' Checks if signature.Parameter corresponds to *args or **kwargs type input.'''
+    if (param.kind == param.KEYWORD_ONLY or param.VAR_KEYWORD) and param.default is param.empty:
```
If you mean to check against both `KEYWORD_ONLY` and `VAR_KEYWORD` you must write:
```python
if param.kind in (param.KEYWORD_ONLY, param.VAR_KEYWORD) and param.default is param.empty:
   ...
```

tommyod commented on treedoc/utils.py:
```diff
@@ -289,6 +289,71 @@ def ispackage(obj):
+    # Check if object has signature
+    try:
+        sig = inspect.signature(obj)
+    except (ValueError, TypeError) as error:
```
You can write:
```python
try:
    sig = inspect.signature(obj)
except ValueError as error:
    pass # action here
except TypeError as error:
    pass # action here
```

tommyod commented on treedoc/utils.py:
```diff
@@ -289,6 +289,71 @@ def ispackage(obj):
+def format_signature(obj, verbosity=2):
+    ''' Format a function signature for printing.'''
+    # TODO: Figure out how to handle *
+    # TODO: What to return if functions have no arguments?
```
Identify the case, and return `()`.

smu095 commented:
I'm failing a test in Travis and I'm not sure why. So, when I test `format_signature` in my terminal, I get the following output:

```bash
>>> from treedoc.utils import format_signature
>>> def myfunc1(a, b: int, *args, c=4.2, d: int=42, **kwargs): return None
... 
>>> format_signature(myfunc1, 4)
'(a, b: int, *args, c=4.2, d: int = 42, **kwargs)'
```
Note the whitespaces that appear to be inserted in the type annotated argument `d`. The error message from Travis reads:

```bash
    def test_signature():
        '''
        Test that formatting signature works.
        '''
    
        def myfunc1(a, b: int, *args, c=4.2, d: int=42, **kwargs):
            return None
    
        assert format_signature(myfunc1, 0) == ""
        assert format_signature(myfunc1, 1) == "(...)"
        assert format_signature(myfunc1, 2) == "(a, b, *args, c, d, **kwargs)"
        assert format_signature(myfunc1, 3) == "(a, b, *args, c=4.2, d=42, **kwargs)"
>       assert format_signature(myfunc1, 4) == "(a, b: int, *args, c=4.2, d: int=42, **kwargs)"
E       AssertionError: assert '(a, b:int, *...42, **kwargs)' == '(a, b: int, *...42, **kwargs)'
E         - (a, b:int, *args, c=4.2, d:int=42, **kwargs)
E         + (a, b: int, *args, c=4.2, d: int=42, **kwargs)
E         ?       +                     +
```
The output of `format_signature` does not appear to be consistent with with what I get in the terminal; here the whitespaces appear to be stripped for all type annotated variables. What am I missing?

smu095 commented:
> That's weird.
> 
> * Is your local Python at 3.7? `python --version`

Jepp, 3.7.3 to be exact.

> * What happens if you ruin the test suite on your computer? See `.travis.yml` for the command to run.

```bash
    def test_signature():
        '''
        Test that formatting signature works.
        '''
    
        def myfunc1(a, b: int, *args, c=4.2, d: int=42, **kwargs):
            return None
    
        assert format_signature(myfunc1, 0) == ""
        assert format_signature(myfunc1, 1) == "(...)"
        assert format_signature(myfunc1, 2) == "(a, b, *args, c, d, **kwargs)"
        assert format_signature(myfunc1, 3) == "(a, b, *args, c=4.2, d=42, **kwargs)"
>       assert format_signature(myfunc1, 4) == "(a, b: int, *args, c=4.2, d: int=42, **kwargs)"
E       AssertionError: assert '(a, b: int, ...42, **kwargs)' == '(a, b: int, *...42, **kwargs)'
E         - (a, b: int, *args, c=4.2, d: int = 42, **kwargs)
E         ?                                 - -
E         + (a, b: int, *args, c=4.2, d: int=42, **kwargs)
```

So on my computer I get a different kind of error.

tommyod commented:
I see two options:
1. Create a utility function `strip_whitespace(string)` and compare the whitespace-stripped versions. That's what's important in essence, and I can resolve this later on. No need for you to face this difficulty on your first PR.
2. Dig into the Signature object and figure out what is happening. You could replace the `str` call by a custom function which extracts the information, or look at the source of `Signature.__str__` to figure out what's going on. 

I suggest you implement 1. to finish up this PR :+1:

tommyod commented:
`tommyod commented 2 hours from now`. Hah.

smu095 commented:
Note to self: It appears that the `Signature.__str__` produces different outputs in 3.6 and 3.7.
